modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTransactionalCache.java

```diff
@@ -627,6 +652,177 @@ else if (!F.isEmpty(req.keys()))
+
+                                req.version(ver);
+                                req.forSavepoint();
+
+                                unlockFuts.put(req, miniFutId);
+                            }
+                        }
+
+                        // Remove candidate from local node first.
+                        GridCacheMvccCandidate rmv = entry.removeLock();
+
+                        if (rmv != null) {
+                            if (!rmv.reentry()) {
+                                if (ver != null && !ver.equals(rmv.version()))
+                                    throw new IgniteCheckedException("Failed to unlock " +
+                                        "(if keys were locked separately, " +
+                                        "then they need to be unlocked separately): " + keys);
+
+                                if (!primary.isLocal()) {
+                                    req.addKey(entry.key(), ctx);
+
+                                    fut.addKey(key, unlockFuts.get(req));
+                                }
+                                else
+                                    locKeys.add(cacheKey);
+
+                                if (log.isDebugEnabled())
+                                    log.debug("Removed lock (will distribute): " + rmv);
+                            }
+                            else if (log.isDebugEnabled())
+                                log.debug("Current thread still owns lock (or there are no other nodes)" +
+                                    " [lock=" + rmv + ", curThreadId=" + Thread.currentThread().getId() + ']');
+                        }
+                    }
+
+                    assert !topVer.equals(AffinityTopologyVersion.NONE) || cand == null;
+
+                    if (topVer.equals(AffinityTopologyVersion.NONE))
+                        topVer = ctx.affinity().affinityTopologyVersion();
+
+                    ctx.evicts().touch(entry, topVer);
+
+                    break;
+                }
+                catch (GridCacheEntryRemovedException ignore) {
+                    if (log.isDebugEnabled())
+                        log.debug("Attempted to unlock removed entry (will retry): " + entry);
+                }
+            }
+        }
+
+        if (ver == null)
+            return;
+
+        for (Map.Entry<ClusterNode, GridNearUnlockRequest> mapping : map.entrySet()) {
+            ClusterNode n = mapping.getKey();
+
+            GridNearUnlockRequest req = mapping.getValue();
+
+            if (n.isLocal())
+                dht.removeLocks(ctx.nodeId(), req.version(), locKeys, true);
+            else if (!F.isEmpty(req.keys())) {
+                try {
+                    ctx.io().send(n, req, ctx.ioPolicy());
+                }
+                catch (ClusterTopologyCheckedException e) {
+                    e.retryReadyFuture(ctx.shared().nextAffinityReadyFuture(tx.topologyVersion()));
```

Comment:
exception is never used ...
may be we should implement exception handling like in org.apache.ignite.internal.processors.cache.distributed.near.GridNearLockFuture#proceedMapping0